server/src/main/java/org/elasticsearch/action/admin/indices/stats/IndicesStatsResponse.java

```diff
@@ -42,10 +48,12 @@
         int totalShards,
         int successfulShards,
         int failedShards,
-        List<DefaultShardOperationFailedException> shardFailures
+        List<DefaultShardOperationFailedException> shardFailures,
+        ClusterState clusterState
     ) {
         super(totalShards, successfulShards, failedShards, shardFailures);
         this.shards = shards;
+        this.clusterState = clusterState;
```

Comment:
I was having hard timing deciding how to approach it as well. That's why this draft contains the least invasive option to start with.

I agree these are the most obvious alternatives, I had the following concerns:

**1) Keep a map with index state and health**
I have to admit this was my preference because it feels like a pretty lean construction to keep this info and it feels like the most logical place. My concern was about the overhead it adds. I am assuming there is a reason we only calculate the indices stats when requested and not in the constructor, in order to find the relevant indices we would have to loop over  the shards table and I was wondering is there any scenario where this would be too heavy?

**2) Add it to the `ShardStats`**
This feels nicer code wise, the new information will be available right where we need it, no need for extra maps etc. My main concern was, if I chose this am I doing it only because it's convenient. What I mean with that is, an index has one or more shards, the `ShardStats` class contains information about a single shard, (if I am not mistaken) adding index information feels like we are stretching the scope of the `ShardStats` to contain also some index information. This feels more "convenient" than "right" to me.

What do you think?

Reply:
I gave it a try and implemented the extra maps because it felt the best option. Re-reading your initial comment, you say:

> it would allow us to expose this information (state and health) in the shard stats API as well, re-using it for this particular use case

This is not covered in this solution. Do you think this benefit is strong enough to justify adding to the stats of a single shards information about the whole index?